core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,198 @@
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    ValidationException.check(updatedRefs.get(name) == null, "Reference with name %s already exists", name);
+    SnapshotRef branch = SnapshotRef.branchBuilder(snapshotId).build();
+    SnapshotRef existingRef = updatedRefs.put(name, branch);
+    ValidationException.check(existingRef == null, "Reference with name %s already exists", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation createTag(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    ValidationException.check(updatedRefs.get(name) == null, "Reference with name %s already exists", name);
+    SnapshotRef tag = SnapshotRef.tagBuilder(snapshotId).build();
+    SnapshotRef existingRef = updatedRefs.put(name, tag);
+    ValidationException.check(existingRef == null, "Reference with name %s already exists", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation removeBranch(String name) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such branch with name %s", name);
+    ValidationException.check(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    updatedRefs.remove(name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation removeTag(String name) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such tag with name %s", name);
+    ValidationException.check(ref.isTag(), "Ref with name %s is a branch not a tag", name);
+    updatedRefs.remove(name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation renameBranch(String name, String newName) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    Preconditions.checkNotNull(newName, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such ref with name %s", name);
+    ValidationException.check(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    updatedRefs.put(newName, ref);
+    updatedRefs.remove(name, ref);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceBranch(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "Branch %s does not exist", name);
+    ValidationException.check(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    SnapshotRef updatedRef = SnapshotRef.builderFrom(ref, snapshotId).build();
+    updatedRefs.put(name, updatedRef);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceTag(String name, long snapshotId) {
```

Comment:
Should this be `updateTag`? Or is it really replacing?

Reply:
I think updateTag is better. I was going off our previous threads where discussed replaceTag, but really what this is doing is just taking an existing tag and updating the snapshot Id it's pointing at.

Reply:
Actually, take back what I said. assign does not really imply the original tag reference changed in terms of where it's pointing. People may interpret it as a create. So I will use updateTag, and replaceBranch